README.rst
=============
Aldryn Search
=============

This package provides a search indexes for easy Haystack 2 integration with django CMS.

Contributing
============

This is a an open-source project. We'll be delighted to receive your
feedback in the form of issues and pull requests. Before submitting your
pull request, please review our `contribution guidelines
<http://docs.django-cms.org/en/latest/contributing/index.html>`_.

We're grateful to all contributors who have helped create and maintain this package.
Contributors are listed at the `contributors <https://github.com/divio/aldryn-search/graphs/contributors>`_
section.

Usage
=====

After installing aldryn-search through your package manager of choice, add ``aldryn_search`` to your
``INSTALLED_APPS``. If you run a multilingual CMS setup, you have to define a haystack backend for every language
in use::

    HAYSTACK_CONNECTIONS = {
        'en': {
            'ENGINE': 'haystack.backends.solr_backend.SolrEngine',
            'URL': 'http://my-solr-server/solr/my-site-en/',
            'TIMEOUT': 60 * 5,
            'INCLUDE_SPELLING': True,
            'BATCH_SIZE': 100,
        },
        'fr': {
            'ENGINE': 'haystack.backends.solr_backend.SolrEngine',
            'URL': 'http://my-solr-server/solr/my-site-fr/',
            'TIMEOUT': 60 * 5,
            'INCLUDE_SPELLING': True,
            'BATCH_SIZE': 100,
        },
    }

To make sure the correct backend is used during search, add ``aldryn_search.router.LanguageRouter`` to your
``HAYSTACK_ROUTERS`` setting::

    HAYSTACK_ROUTERS = ['aldryn_search.router.LanguageRouter',]



When using multiple languages, usually there's one search backend per language, when indexing it's important to know
which language is currently being used, this can be facilitated by the ``ALDRYN_SEARCH_LANGUAGE_FROM_ALIAS`` setting,
this setting could be a callable or a string path that resolves to one.

Please keep in mind that it's usually not a good idea to import things in your settings, however there are cases where
it seems overkill to create a function to handle the alias, for example::

    ALDRYN_SEARCH_LANGUAGE_FROM_ALIAS = lambda alias: alias.split('-')[-1]


the example above could be used when using multiple languages and sites, all backends could have a language suffix.

The same could be achieved using a function defined somewhere else in your code like so::

    ALDRYN_SEARCH_LANGUAGE_FROM_ALIAS = "my_app.helpers.language_from_alias"



If any of the above return None then ``settings.LANGUAGE_CODE`` will be used.

By default this setting evaluates to a function that checks if the alias is in ``settings.LANGUAGES`` and if so it
uses the alias as a language.


For a complete Haystack setup, please refer to their `documentation <https://django-haystack.readthedocs.io/>`_.

For more documentation, see the `docs folder <https://github.com/divio/aldryn-search/tree/master/docs/>`_.

Integration with django CMS
===========================

aldryn-search comes with an App Hook for django CMS, and a search view using Django's class based views. If you
want to use this app hook, you can either subclass it and register it yourself, or set
``ALDRYN_SEARCH_REGISTER_APPHOOK`` to ``True``.

If you want to exclude some cms plugins from indexing, you can specify ``ALDRYN_SEARCH_PLUGINS_EXCLUDE`` setting like so::

    ALDRYN_SEARCH_EXCLUDED_PLUGINS = [
        "PluginA", "PluginB"
    ]

For pagination, aldryn-search uses the ``aldryn_search.paginators.DiggPaginator``. If you want to use this built-in
pagination, make sure to install `django-spurl <https://github.com/j4mie/django-spurl>`_, and add then add ``spurl``
to ``INSTALLED_APPS``.

Pagination
==========

Results are paginated according to the ``ALDRYN_SEARCH_PAGINATION`` setting (default: 10).
If set to ``None`` pagination is disabled.


CHANGELOG.txt
=========
Changelog
=========


1.1.1 (2021-02-12)
==================
* Add ``DiggPaginator`` from deprecated ``aldryn_common`` library to codebase
* Remove ``aldryn_common`` from dependencies


1.1.0 (2020-02-20)
==================

* Added support for Django 2.2 and django CMS 3.7
* Added support for Transifex
* Added isort and flake8
* Changed from djangocms-helper to django-app-helper
* Extended test matrix
* Adapted code base to align with other supported addons
* Removed ``INDEX_TITLE`` backwards compatibility shim


1.0.1 (2018-06-14)
==================

* Added support for configurable alias from language function
  with the new ``ALDRYN_SEARCH_ALIAS_FROM_LANGUAGE`` setting


1.0.0 (2018-04-05)
==================

* Added django CMS 3.5 to test matrix
* Dropped support for Django < 1.8
* Dropped suppport for django CMS < 3.2


0.5.0 (2018-03-06)
==================

* Added new ``ALDRYN_SEARCH_EXCLUDED_PLUGINS`` setting to prevent
  specific plugin types from being indexed


0.4.1 (2017-11-02)
==================

* Updated metadata for pypi


0.4.0 (2017-10-20)
==================

* Added CMS 3.5.x compatibility
* Dropped support for Django 1.7.x


0.3.0 (2016-10-20)
==================

* Added CMS 3.4.x compatibility
* Added Django 1.8.x compatibility


0.2.12 (2016-05-19)
==================

* Added CMS 3.3.x compatibility


0.2.11 (2016-02-03)
==================

* Added Django 1.9 compatibility
* Dropped support for django CMS 2.4
* Fixed Travis setup
* Fixed importlib deprecation warning
* Fixed bug where unpublishing cms page would not remove it from index
* Added support for explicit signal based indexing
* Fixed regression on plugin indexing
* Restored the old _strip_tags due to a regression


0.2.10 (2015-10-15)
==================

* Updated documentation
* Added Django 1.8 compatibility


0.2.9 (2015-08-05)
==================

* Tests
* Fixed issue with lxml parser error


0.2.8 (2015-07-06)
==================

* Fixed issues with strip_tags
* Fixed issue where the non-public version of a title was being indexed


0.2.7 (2015-04-16)
==================

* Fixed a issue with get_language return wrong language in indexing


0.2.6 (2015-01-27)
==================

* Fixed issue with search_fields not following relationships


0.2.5 (2015-01-27)
==================

* Fixed issue with javascript not being stripped correctly
* Fixed duplicate content issue when indexing a plugin that set search_fields


0.2.4 (2014-11-24)
==================

* Added missing requirements to setup.py (issue #23)
* Added for_write router. Fixes issues on single object updates


0.2.3 (2014-09-30)
==================

* Fixed TypeError when rebuilding page index on cms 2.4
* Fixed typo on README
* Fixed a bug on spelling suggestion check when using pagination


0.2.2 (2014-08-06)
==================

* Fixed bug in get_model_path utility (issue #19)


0.2.1 (2014-07-25)
==================

* Fixed bug in form handling on search view


0.2.0 (2014-07-24)
==================

* Removed contrib.pagination.DiggPaginator (use aldryn_common.paginator.DiggPaginator)
* Added alias_from_language utility function
* Renamed searchqueryset to search_queryset in search view
* Added search_queryset_class to search view

* Added helpers module
* Added helper function get_plugin_index_data
* Added new utility clean_join
* Added models attribute on view, to limit search by specific models
* Renamed get_request_for_search to get_request and moved it to helpers module
* Added get_request_instance method to base index, this allows adding custom attributes to request object
* Renamed INDEX_TITLE to index_title to keep the same format as haystack
* Refactored plugin data indexing to split on whitespace, thus reducing index size


0.1.9 (2014-05-17)
0.1.8 (2014-05-16)
0.1.7 (2014-03-20)
0.1.6 (2014-01-30)
0.1.5 (2014-01-29)
0.1.4 (2014-01-29)
0.1.3 (2013-11-11)
0.1.2 (2013-08-07)


aldryn_search/helpers.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.contrib.auth.models import AnonymousUser
from django.template import Engine, RequestContext
from django.test import RequestFactory
from django.utils.text import smart_split

from cms.toolbar.toolbar import CMSToolbar

from .conf import settings
from .utils import (
    _get_alias_from_language_func, _get_language_from_alias_func,
    get_field_value, strip_tags,
)


try:
    from django.utils.encoding import force_unicode
except ImportError:
    from django.utils.encoding import force_text as force_unicode


EXCLUDED_PLUGINS = getattr(settings, 'ALDRYN_SEARCH_EXCLUDED_PLUGINS', [])


get_alias_from_language = _get_alias_from_language_func()
get_language_from_alias = _get_language_from_alias_func()


def _render_plugin(plugin, context, renderer=None):
    if renderer:
        content = renderer.render_plugin(
            instance=plugin,
            context=context,
            editable=False,
        )
    else:
        content = plugin.render_plugin(context)
    return content


def get_cleaned_bits(data):
    decoded = force_unicode(data)
    stripped = strip_tags(decoded)
    return smart_split(stripped)


def get_plugin_index_data(base_plugin, request):
    text_bits = []

    instance, plugin_type = base_plugin.get_plugin_instance()

    if instance is None or instance.plugin_type in EXCLUDED_PLUGINS:
        # this is an empty plugin or excluded from search
        return text_bits

    search_fields = getattr(instance, 'search_fields', [])
    if hasattr(instance, 'search_fulltext'):
        # check if the plugin instance has search enabled
        search_contents = instance.search_fulltext
    elif hasattr(base_plugin, 'search_fulltext'):
        # now check in the base plugin instance (CMSPlugin)
        search_contents = base_plugin.search_fulltext
    elif hasattr(plugin_type, 'search_fulltext'):
        # last check in the plugin class (CMSPluginBase)
        search_contents = plugin_type.search_fulltext
    else:
        # disabled if there's search fields defined,
        # otherwise it's enabled.
        search_contents = not bool(search_fields)

    if search_contents:
        context = RequestContext(request)
        updates = {}
        engine = Engine.get_default()

        for processor in engine.template_context_processors:
            updates.update(processor(context.request))
        context.dicts[context._processors_index] = updates

        try:
            # django-cms>=3.5
            renderer = request.toolbar.content_renderer
        except AttributeError:
            # django-cms>=3.4
            renderer = context.get('cms_content_renderer')

        plugin_contents = _render_plugin(instance, context, renderer)

        if plugin_contents:
            text_bits = get_cleaned_bits(plugin_contents)
    else:
        values = (get_field_value(instance, field) for field in search_fields)

        for value in values:
            cleaned_bits = get_cleaned_bits(value or '')
            text_bits.extend(cleaned_bits)
    return text_bits


def get_request(language=None):
    """
    Returns a Request instance populated with cms specific attributes.
    """
    request_factory = RequestFactory(HTTP_HOST=settings.ALLOWED_HOSTS[0])
    request = request_factory.get("/")
    request.session = {}
    request.LANGUAGE_CODE = language or settings.LANGUAGE_CODE
    # Needed for plugin rendering.
    request.current_page = None
    request.user = AnonymousUser()
    request.toolbar = CMSToolbar(request)
    return request
